runtime/near-evm-runner/src/builtins.rs

```diff
@@ -0,0 +1,587 @@
+            if let Ok(p) = secp256k1::recover(&hash, &s, &rec_id) {
+                // recover returns the 65-byte key, but addresses come from the raw 64-byte key
+                let r = sha3::Keccak256::digest(&p.serialize()[1..]);
+                output.write(0, &[0; 12]);
+                output.write(12, &r[12..]);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Impl for Sha256 {
+    fn execute(&self, input: &[u8], output: &mut BytesRef) -> Result<(), Error> {
+        use sha2::Digest;
+        let d = sha2::Sha256::digest(input);
+        output.write(0, &*d);
+        Ok(())
+    }
+}
+
+impl Impl for Ripemd160 {
+    fn execute(&self, input: &[u8], output: &mut BytesRef) -> Result<(), Error> {
+        let hash = ripemd160::Ripemd160::digest(input);
+        output.write(0, &[0; 12][..]);
+        output.write(12, &hash);
+        Ok(())
+    }
+}
+
+// calculate modexp: left-to-right binary exponentiation to keep multiplicands lower
+fn modexp(mut base: BigUint, exp: Vec<u8>, modulus: BigUint) -> BigUint {
+    const BITS_PER_DIGIT: usize = 8;
+
+    // n^m % 0 || n^m % 1
+    if modulus <= BigUint::one() {
+        return BigUint::zero();
+    }
+
+    // normalize exponent
+    let mut exp = exp.into_iter().skip_while(|d| *d == 0).peekable();
+
+    // n^0 % m
+    if exp.peek().is_none() {
+        return BigUint::one();
+    }
+
+    // 0^n % m, n > 0
+    if base.is_zero() {
+        return BigUint::zero();
+    }
+
+    base %= &modulus;
+
+    // Fast path for base divisible by modulus.
+    if base.is_zero() {
+        return BigUint::zero();
+    }
+
+    // Left-to-right binary exponentiation (Handbook of Applied Cryptography - Algorithm 14.79).
+    // http://www.cacr.math.uwaterloo.ca/hac/about/chap14.pdf
+    let mut result = BigUint::one();
+
+    for digit in exp {
+        let mut mask = 1 << (BITS_PER_DIGIT - 1);
+
+        for _ in 0..BITS_PER_DIGIT {
+            result = &result * &result % &modulus;
+
+            if digit & mask > 0 {
+                result = result * &base % &modulus;
+            }
+
+            mask >>= 1;
+        }
+    }
+
+    result
+}
+
+impl Impl for ModexpImpl {
+    fn execute(&self, input: &[u8], output: &mut BytesRef) -> Result<(), Error> {
+        let mut reader = input.chain(io::repeat(0));
+        let mut buf = [0; 32];
+
+        // read lengths as usize.
+        // ignoring the first 24 bytes might technically lead us to fall out of consensus,
+        // but so would running out of addressable memory!
+        let mut read_len = |reader: &mut io::Chain<&[u8], io::Repeat>| {
+            reader
+                .read_exact(&mut buf[..])
+                .expect("reading from zero-extended memory cannot fail; qed");
+            BigEndian::read_u64(&buf[24..]) as usize
+        };
+
+        let base_len = read_len(&mut reader);
+        let exp_len = read_len(&mut reader);
+        let mod_len = read_len(&mut reader);
+
```

Comment:
Let's use gas for this, instead of adhock limits.

https://github.com/nearprotocol/nearcore/blob/b7b820abe7babc043ae6dc5087e40f097347d2ea/runtime/near-evm-runner/src/builtins.rs#L231